zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,556 @@
+			struct vmh_active_allocation *active_alloc = CONTAINER_OF(
+				iterator_activ, struct vmh_active_allocation, node);
+
+			if (vmh_is_range_in_range(
+				active_alloc->allocation_ptr,
+				active_alloc->allocation_size,
+				physical_maping->mapping_ptr,
+				physical_maping->mapping_size)) {
+
+				can_free = false;
+				break;
+			}
+		}
+		if (can_free) {
+			sys_mm_drv_unmap_region(physical_maping->mapping_ptr,
+				physical_maping->mapping_size);
+			list_item_del(iterator_phys);
+			break;
+		}
+	}
+
+	rfree(physical_maping);
+	rfree(allocation);
+
+	return retval;
+}
+
+/**
+ * @brief Reconfigure heap with given config
+ *
+ * This will destroy the heap from the pointer and recreate it
+ * using provided config. Individual region attribute is the
+ * "anchor" to the virtual space we use.
+ *
+ * @param heap Ptr to the heap that should be reconfigured
+ * @param cfg heap blocks configuration
+ * @param span allow contiguous on the reconfigured heap
+ *
+ * @retval heap_pointer Returns new heap pointer on success
+ * @retval NULL when reconfiguration failed
+ */
+struct virtual_memory_heap *vmh_reconfigure_heap(
+	struct virtual_memory_heap *heap,
+	struct heap_config *cfg, bool span)
+{
+	uint32_t region_attribute = heap->virtual_region->attr;
+
+	if (vmh_free_heap(heap))
+		return NULL;
+
+	return vmh_init_heap(cfg, region_attribute, span);
+}
+
+/**
+ * @brief Get default configuration for heap
+ *
+ * This will return config created based on region provided.
+ * It will try to split memory in even chunks and split it into
+ * bundles of size 64,128,...,1024 - exactly 5 sizes.
+ *
+ * @param region Virtual region which resources we will be using for heap
+ *
+ * @retval Returns default heap_config
+ */
+void vmh_get_default_heap_config(
+	const struct sys_mm_drv_region *region,
+	struct heap_config *cfg)
+{
+	int buffer_size = region->size / 5;
+	int block_count = buffer_size / DCACHE_LINE_SIZE;
+
+	for (size_t i = 0; i < 5; i++) {
+		cfg->block_bundles_table[i].block_size =
+			DCACHE_LINE_SIZE << i;
+		cfg->block_bundles_table[i].number_of_blocks =
+			buffer_size / cfg->block_bundles_table[i].block_size;
+	}
+}
+
+/**
+ * @brief Initialization of static objects in virtual heaps api
+ *
+ * This will initialize lists of allocations and physical mappings.
+ *
+ * @param unused Variable is unused - needed by arch to clear warning.
+ *
+ * @retval Returns 0 on success
+ */
+static int virtual_heaps_init(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	list_init(&vm_heaps);
+	list_init(&physical_mappings);
 	return 0;
 }
 
 SYS_INIT(virtual_heaps_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
+
+/**
+ * @brief Gets pointer to already created heap with identified by
+ * provided attribute.
+ *
+ * @param attr Virtual memory region attribute.
+ *
+ * @retval heap ptr on success
+ * @retval NULL if there was no heap created fitting the attr.
+ */
+struct virtual_memory_heap *vmh_get_heap_by_attribute(uint32_t attr)
```

Comment:
I don't exactly understand how this works.
It looks like this function is capable of finding only 1 instance of the heap matching attribute (returns first one found).
But `vmh_init_heap` function does not restrict us from creating many heap instances lets say on given core (with MEM_REG_ATTR_CORE_HEAP). What is the proper solution to this problem? If we restrict by implementation that a single core may have only 1 heap, we should restrict this in init. If we don't, this function should probably return a collection of heaps with matching attribute right?
In short - it looks like attribute does not uniquely identify the specific heap instance. Or am I wrong?

Reply:
goood catch - heaps are unique and only core heaps have more instances depending on number of cores fixed in next code iteration